runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -3396,6 +3198,551 @@ J9::Z::TreeEvaluator::ArrayCopyBNDCHKEvaluator(TR::Node * node, TR::CodeGenerato
+      TR::Register *valueReferenceReg = cg->evaluate(valueNode);
+
+      TR::DataType nodeType = valueNode->getDataType();
+      TR::SymbolReference *sr = cg->allocateLocalTemp(nodeType);
+      TR::MemoryReference *valueMR = generateS390MemoryReference(valueNode, sr, cg);
+      if (valueReferenceReg->getKind() == TR_GPR)
+         {
+         // Use STG if the dataType is an uncompressed TR::Address or TR::Int64. ST otherwise.
+         auto mnemonic = TR::DataType::getSize(nodeType) == 8 ? TR::InstOpCode::STG : TR::InstOpCode::ST;
+         // Now store the value onto the stack.
+         generateRXInstruction(cg, mnemonic, node, valueReferenceReg, valueMR);
+         }
+      else if (valueReferenceReg->getKind() == TR_FPR)
+         {
+         auto mnemonic = nodeType == TR::Float ? TR::InstOpCode::STE : TR::InstOpCode::STD;
+         // Now store the value onto the stack.
+         generateRXInstruction(cg, mnemonic, node, valueReferenceReg, valueMR);
+         }
+      else
+         {
+         TR_ASSERT_FATAL(false, "Unsupported register kind (%d) for fieldwatch.", valueReferenceReg->getKind());
+         }
+      valueReferenceReg = cg->allocateRegister();
+
+      // Now load the memory location back into the register so that it can be used
+      // as an argument register for the VM helper call.
+      TR::MemoryReference *tempMR = generateS390MemoryReference(*valueMR, 0, cg);
+      generateRXInstruction(cg, TR::InstOpCode::LA, node, valueReferenceReg, tempMR);
+
+      dependencies->addPreCondition(valueReferenceReg, helperLink->getIntegerArgumentRegister(numArgs));
+      dependencies->addPostCondition(valueReferenceReg, helperLink->getIntegerArgumentRegister(numArgs));
+      valueReferenceReg->setPlaceholderReg();
+
+      cg->stopUsingRegister(valueReferenceReg);
+      }
+
+   // These registers will hold Entry Point and Return Address registers, which are required when generating a directCall.
+   TR::Register *scratch1 = cg->allocateRegister();
+   TR::Register *scratch2 = cg->allocateRegister();
+   dependencies->addPostCondition(scratch1, cg->getEntryPointRegister());
+   dependencies->addPostCondition(scratch2, cg->getReturnAddressRegister());
+
+   // Now generate the call to VM Helper to report the fieldwatch.
+   TR::Instruction *call = generateDirectCall(cg, node, false /*myself*/, cg->symRefTab()->findOrCreateRuntimeHelper(helperIndex, false, false, false), dependencies);
+   call->setNeedsGCMap(0x0000FFFF);
+   call->setDependencyConditions(dependencies);
+   cg->resetIsLeafMethod();
+
+   // After returning from the VM Helper, branch back to mainline code.
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, mergePointLabel);
+
+   cg->stopUsingRegister(scratch1);
+   cg->stopUsingRegister(scratch2);
+
+   // TODO: this may be unnecessary
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_MASK15, node, endLabel);
```

Comment:
It is indeed unnecessary. We should remove this.

Reply:
I'm not sure how this keeps escaping my cherry-picks as I've removed this multiple times and it doesn't exist in my test branch. (I'm cherry-picking changes from my test branch to this branch once I'm confident that they work).

Anyway, I've done some re-syncing here: 0dcb940